vibrato/src/trainer_config.rs

```diff
@@ -0,0 +1,280 @@
+                    ));
+                }
+            } else {
+                return Err(VibratoError::invalid_format("feature", ""));
+            }
+        }
+
+        Ok(FeatureExtractor::new(&unigram_templates, &bigram_templates))
+    }
+
+    fn parse_rewrite_rule(line: &str) -> Result<(Vec<&str>, Vec<&str>)> {
+        let mut spl = line.split_ascii_whitespace();
+        let pattern = spl.next();
+        let rewrite = spl.next();
+        let rest = spl.next();
+        if let (Some(pattern), Some(rewrite), None) = (pattern, rewrite, rest) {
+            Ok((pattern.split(',').collect(), rewrite.split(',').collect()))
+        } else {
+            Err(VibratoError::invalid_format(
+                "rewrite.def",
+                "invalid rewrite rule",
+            ))
+        }
+    }
+
+    fn parse_rewrite_config<R>(
+        rdr: R,
+    ) -> Result<(FeatureRewriter, FeatureRewriter, FeatureRewriter)>
+    where
+        R: Read,
+    {
+        let reader = BufReader::new(rdr);
+
+        let mut unigram_rewriter_builder = FeatureRewriterBuilder::new();
+        let mut left_rewriter_builder = FeatureRewriterBuilder::new();
+        let mut right_rewriter_builder = FeatureRewriterBuilder::new();
+
+        let mut builder = None;
+        for line in reader.lines() {
+            let line = line?;
+            let line = line.trim();
+
+            if line.is_empty() || line.starts_with('#') {
+                continue;
+            }
+
+            match line {
+                "[unigram rewrite]" => builder = Some(&mut unigram_rewriter_builder),
+                "[left rewrite]" => builder = Some(&mut left_rewriter_builder),
+                "[right rewrite]" => builder = Some(&mut right_rewriter_builder),
+                line => {
+                    if let Some(builder) = builder.as_mut() {
+                        let (pattern, rewrite) = Self::parse_rewrite_rule(line)?;
+                        builder.add_rule(&pattern, &rewrite)?;
+                    } else {
+                        return Err(VibratoError::invalid_format(
+                            "rewrite.def",
+                            "Invalid rewrite rule",
+                        ));
+                    }
+                }
+            }
+        }
+
+        Ok((
+            FeatureRewriter::from(unigram_rewriter_builder),
+            FeatureRewriter::from(left_rewriter_builder),
+            FeatureRewriter::from(right_rewriter_builder),
+        ))
+    }
+
+    /// Loads a training configuration from readers.
+    ///
+    /// # Arguments
+    ///
+    /// * `feature_templates_rdr` - A reader of the feature definition file `feature.def`.
+    /// * `rewrite_rules_rdr` - A reader of the rewrite definition file `rewrite.def`.
+    /// * `char_prop_rdr` - A reader of the character definition file `char.def`.
+    ///
+    /// # Errors
+    ///
+    /// An error variant returns if the file format is invalid.
```

Comment:
```suggestion
    /// * `feature_templates_rdr`: A reader of the feature definition file `feature.def`.
    /// * `rewrite_rules_rdr`: A reader of the rewrite definition file `rewrite.def`.
    /// * `char_prop_rdr`: A reader of the character definition file `char.def`.
    ///
    /// # Errors
    ///
    /// [`VibratoError`] is returned when an input format is invalid.
    ///
    /// # Notes
    ///
    /// The readers are buffered automatically, so you should not
    /// wrap them in a buffered reader like [`std::io::BufReader`].
```
For consistency with other readers.

Reply:
The official documentation uses `* arg_name - description` style. Shouldn't it follow the official documentation?
https://doc.rust-lang.org/rust-by-example/meta/doc.html

Reply:
I think that some functions should take `BufRead` instead of `Read`.
What do you think?

Reply:
> Shouldn't it follow the official documentation?

Okay. We will follow the official manner. The other parts will be modified in that manner.

> I think that some functions should take BufRead instead of Read.

I want to keep `Read` because:

- It is less confusing to handle `Read` through all arguments.
- Since this library takes multiple readers as arguments, requiring `BufRead` outside the API involves a lot of coding for users (and us in development).

Reply:
Trait boundaries should be sufficient conditions for the function to work best and should not be referred to as "you should not..." through the comment.

In other words, if the trait boundary is `Read`, then the function should work best for structures that implement `Read`, including `BufReader`.
If `Read` is not sufficient, the function should require `BufRead`.

Reply:
In my opinion, the following options would be best:
1. changes the trait to `BufRead` and removes wrapping the argument by `BufReader`.
2. keeps `Read` and removes the Notes section.

Reply:
Thanks for your suggestions.

> In other words, if the trait boundary is Read, then the function should work best for any structure that implements Read, including BufReader. If Read is not sufficient, the function should require BufRead.

Yeah, I agree with taking `BufRead` as an argument if it performs with `BufRead`, because this is the role of the trait boundary. But unfortunately, it is not possible to completely replace all arguments with `BufRead` because of its dependence on `csv` crate. For example, in [`Dictionary::from_readers`](https://docs.rs/vibrato/latest/vibrato/dictionary/struct.Dictionary.html#method.from_readers), `C` and `P` can be replaced with `BufRead`, but `S` and `U` cannot be.

I fear that handling multiple similar `Trait` in the same function may be a bit confusing.

> 2. keeps Read and removes the Notes section.

In my opinion, your second suggestion is the best in the current version. However, if we were to remove the dependency on the `csv` crate, all arguments for readers should be replaced with `BufRead`.